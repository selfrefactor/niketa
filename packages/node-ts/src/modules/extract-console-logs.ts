import {anyFalse, map, remove, take, trim, forEach} from 'rambdax'
const LIMIT = 115
const MARK = 'NIKETA_MARKER'

interface Decoration {
  line: string,
  decoration: string,
}

function mergeLogs(hash: Record<string, any>) {
  const iterable = (x: Record<string, any>) => {
    const lineDecoration: Decoration[] = []

    forEach((xInstance, lineNumber) => {
      const decoration =
        xInstance.length === 1 ? xInstance[0] : xInstance.join(' ')

      lineDecoration.push({
        decoration,
        line: lineNumber,
      })
    }, x)

    return lineDecoration
  }

  return map<any, any>(iterable, hash)
}

export function extractConsoleLogs(input: string) {
  const withMarker = input.replace(/console\.log/g, `${MARK} console.log`)

  const parts = withMarker.split('console.log')
  const hash: Record<string, any> = {}

  const extractor = (part: string) => {
    const partialLines = part
      .split('\n')
      .filter(Boolean)
      .filter(x => !x.includes(MARK))
      .map(trim)

    if (partialLines.length === 0) return

    const logLines: string[] = []
    let found = false
    let fileName = ''
    let lineNumber = 0

    partialLines.forEach(partialLine => {
      if (found) return
      const matched = partialLine.match(
        /\([a-zA-Z_\-\.\/]+:[0-9]+:[0-9]+\)$/
      )

      if (matched === null && !found) return logLines.push(partialLine)
      if (matched === null) return
      found = true

      const [fileNameRaw, lineNumberRaw] = matched[0].split(':')

      lineNumber = Number(lineNumberRaw)
      fileName = remove(['(', ')'])(fileNameRaw)
    })

    if (anyFalse(found, lineNumber, fileName)) return

    if (hash[fileName] === undefined) hash[fileName] = {}
    if (hash[fileName][lineNumber] === undefined)
      hash[fileName][lineNumber] = []

    const decoration = take(LIMIT)(logLines.join(' '))
    hash[fileName][lineNumber].push(decoration)
  }

  parts.forEach(extractor)

  return mergeLogs(hash)
}
