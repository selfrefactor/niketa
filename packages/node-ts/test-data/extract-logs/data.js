export const testInput =
  'console.log src/_modules/parseBeforeNotify.spec.js:10\n' +
  '    string\n' +
  '\n' +
  '  console.log src/_modules/parseBeforeNotify.spec.js:12\n' +
  '    string [ 1, 23, { a: 1 } ]\n' +
  '\n'

export const testInputSecond = `
console.log source/adjust.js:6
1 [ 0, 1, 2 ]

console.log source/adjust.js:6
1 [ 0, 1, 2 ]

console.log source/adjust.js:6
1 [ 0, 1, 2 ]

console.log source/adjust.js:6
1 [ 0, 1, 2 ]

console.log source/adjust.js:6
-2 [ 0, 1, 2 ]

console.log source/adjust.js:6
4 [ 0, 1, 2, 3 ]

console.log source/adjust.js:6
-5 [ 0, 1, 2, 3 ]

-----------|---------|----------|---------|---------|-------------------
File       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
-----------|---------|----------|---------|---------|-------------------
All files  |     100 |      100 |     100 |     100 |                   
adjust.js |     100 |      100 |     100 |     100 |                   
-----------|---------|----------|---------|---------|-------------------
`.trim()

export const testInputThird = `
console.log
object

  at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:5:11)

----------------------|---------|----------|---------|---------|-------------------
File                  | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------------------|---------|----------|---------|---------|-------------------
All files             |   93.75 |       60 |     100 |   93.33 |                   
sort-package-json.js |   93.75 |       60 |     100 |   93.33 | 29                
----------------------|---------|----------|---------|---------|-------------------

`.trim()
