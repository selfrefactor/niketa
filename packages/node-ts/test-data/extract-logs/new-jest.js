export const newJest = `  console.log
{ type: 'object' }

  at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:6:11)

console.log
{ array: 'not array' }

  at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:7:11)

console.log
{ keys: [ 'unsortedKeys', 'sortedKeys' ] }

  at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:8:11)

console.log
{
  unsortedKeys: [
    'name',
    'version',
    'description',
    'author',
    'license',
    'scripts',
    'dependencies',
    'devDependencies',
    'depFn',
    'jest',
    'commitMessage'
  ],
  sortedKeys: [
    'name',
    'scripts',
    'author',
    'version',
    'description',
    'license',
    'dependencies',
    'devDependencies',
    'depFn',
    'jest',
    'commitMessage'
  ]
}

  at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:9:11)

console.log
  { type: 'boolean' }
  
  at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:6:11)

----------------------|---------|----------|---------|---------|-------------------
File                  | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------------------|---------|----------|---------|---------|-------------------
All files             |   92.31 |       60 |     100 |   91.67 |                   
sort-package-json.js |   92.31 |       60 |     100 |   91.67 | 29                
----------------------|---------|----------|---------|---------|-------------------
`

export const bug = `console.log
1

  at Object.<anonymous> (src/_modules/format-json.spec.js:150:10)

console.log
[ 1, 2, 3 ]

  at Object.<anonymous> (src/_modules/format-json.spec.js:151:10)

console.log
{ a: 1 }

  at Object.<anonymous> (src/_modules/format-json.spec.js:152:10)

----------------|---------|----------|---------|---------|-------------------
File            | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------------|---------|----------|---------|---------|-------------------
All files       |   83.33 |      100 |     100 |   83.33 |                   
format-json.js |   83.33 |      100 |     100 |   83.33 | 10                
----------------|---------|----------|---------|---------|-------------------`
