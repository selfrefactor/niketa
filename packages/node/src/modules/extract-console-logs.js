// import { outputFile } from 'fs-extra'
import { anyFalse, map, remove, take, trim } from 'rambdax'
const LIMIT = 115
const MARK = 'NIKETA_MARKER'

function mergeLogs(hash){
  const iterable = x => {
    const lineDecoration = []

    map((xInstance, lineNumber) => {
      const decoration =
        xInstance.length === 1 ? xInstance[ 0 ] : xInstance.join(' ')
      lineDecoration.push({
        decoration,
        line : lineNumber,
      })
    })(x)

    return lineDecoration
  }

  return map(iterable)(hash)
}

export function extractConsoleLogs(input){
  // outputFile(`${ __dirname }/latest-log.txt`, input)
  const withMarker = input.replace(/console\.log/g, `${ MARK } console.log`)

  const parts = withMarker.split('console.log')
  const hash = {}

  const extractor = part => {
    const partialLines = part
      .split('\n')
      .filter(Boolean)
      .filter(x => !x.includes(MARK))
      .map(trim)

    if (partialLines.length === 0) return

    const logLines = []
    let found = false
    let fileName = ''
    let lineNumber = 0

    partialLines.forEach(partialLine => {
      if (found) return
      const matched = partialLine.match(/\([a-zA-Z_\-\.\/]+:[0-9]+:[0-9]+\)$/)

      if (!matched && !found) return logLines.push(partialLine)
      found = true

      const [ fileNameRaw, lineNumberRaw ] = matched[ 0 ].split(':')

      lineNumber = Number(lineNumberRaw)
      fileName = remove([ '(', ')' ])(fileNameRaw)
    })
    if (anyFalse(
      found, lineNumber, fileName
    )) return

    if (hash[ fileName ] === undefined) hash[ fileName ] = {}
    if (hash[ fileName ][ lineNumber ] === undefined)
      hash[ fileName ][ lineNumber ] = []

    const decoration = take(LIMIT)(logLines.join(' '))
    hash[ fileName ][ lineNumber ].push(decoration)
  }

  parts.forEach(extractor)

  return mergeLogs(hash)
}
