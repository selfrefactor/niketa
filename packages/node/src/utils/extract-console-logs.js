import { take } from 'rambdax'
const LIMIT = 60

function mergeLogs(logData){
  const toReturn = {}

  Object.keys(logData).forEach(key => {
    const currentDecoration = logData[ key ]
    const foundLines = []
    const mergedDecoration = []

    currentDecoration.forEach(({ line, decoration }, i) => {
      if (foundLines.includes(line)) return
      foundLines.push(line)

      const allLogs = currentDecoration
        .filter(x => x.line === line)
        .map(x => x.decoration)

      if (allLogs.length === 1){
        return mergedDecoration.push({
          line,
          decoration,
        })
      }
      const joinedDecoration = take(LIMIT, allLogs.join(', '))

      return mergedDecoration.push({
        line,
        decoration : joinedDecoration,
      })
    })
    toReturn[ key ] = mergedDecoration
  })

  return toReturn
}

export function extractConsoleLogs(input){
  const lines = input.split('\n')
  console.log(1)
  console.log({
    lines,
    len : lines.length,
  }) 
  const toReturn = {}

  lines.forEach((line, i) => {
    if (line.trim().startsWith('console.log')){
      const [ , fileNameAndLine ] = line.split('console.log')
      console.log(12, {
        line,
        fileNameAndLine,
      })
      if (!fileNameAndLine.includes(':')) return

      const [ fileName, lineNumber ] = fileNameAndLine
        .split(':')
        .map(x => x.trim())
      if (toReturn[ fileName ] === undefined) toReturn[ fileName ] = []
      const log = lines[ i + 1 ]
      if (log === undefined) return

      toReturn[ fileName ].push({
        line       : lineNumber,
        decoration : log.trim(),
      })
    }
  })

  return mergeLogs(toReturn)
}
