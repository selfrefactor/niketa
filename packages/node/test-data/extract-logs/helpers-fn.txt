 console.log
    { type: 'object' }

      at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:5:11)

  console.log
    { array: 'not array' }

      at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:6:11)

  console.log
    { keys: [ 'unsortedKeys', 'sortedKeys' ] }

      at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:7:11)

  console.log
    {
      unsortedKeys: [
        'name',
        'version',
        'description',
        'author',
        'license',
        'scripts',
        'dependencies',
        'devDependencies',
        'depFn',
        'jest',
        'commitMessage'
      ],
      sortedKeys: [
        'name',
        'scripts',
        'author',
        'version',
        'description',
        'license',
        'dependencies',
        'devDependencies',
        'depFn',
        'jest',
        'commitMessage'
      ]
    }

      at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:8:11)

----------------------|---------|----------|---------|---------|-------------------
File                  | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------------------|---------|----------|---------|---------|-------------------
All files             |   93.75 |       60 |     100 |   93.33 |                   
 sort-package-json.js |   93.75 |       60 |     100 |   93.33 | 29                
----------------------|---------|----------|---------|---------|-------------------
