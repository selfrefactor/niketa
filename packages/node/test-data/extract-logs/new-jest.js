export const newJest = `  console.log
{ type: 'object' }

  at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:6:11)

console.log
{ array: 'not array' }

  at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:7:11)

console.log
{ keys: [ 'unsortedKeys', 'sortedKeys' ] }

  at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:8:11)

console.log
{
  unsortedKeys: [
    'name',
    'version',
    'description',
    'author',
    'license',
    'scripts',
    'dependencies',
    'devDependencies',
    'depFn',
    'jest',
    'commitMessage'
  ],
  sortedKeys: [
    'name',
    'scripts',
    'author',
    'version',
    'description',
    'license',
    'dependencies',
    'devDependencies',
    'depFn',
    'jest',
    'commitMessage'
  ]
}

  at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:9:11)

console.log
  { type: 'boolean' }
  
  at Object.<anonymous> (libs/sort-package-json/sort-package-json.spec.js:6:11)

----------------------|---------|----------|---------|---------|-------------------
File                  | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------------------|---------|----------|---------|---------|-------------------
All files             |   92.31 |       60 |     100 |   91.67 |                   
sort-package-json.js |   92.31 |       60 |     100 |   91.67 | 29                
----------------------|---------|----------|---------|---------|-------------------
`
