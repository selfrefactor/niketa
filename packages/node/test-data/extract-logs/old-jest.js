export const oldJest = `  console.log libs/sort-package-json/sort-package-json.spec.js:6
{ type: 'object' }

console.log libs/sort-package-json/sort-package-json.spec.js:7
{ array: 'not array' }

console.log libs/sort-package-json/sort-package-json.spec.js:8
{ keys: [ 'unsortedKeys', 'sortedKeys' ] }

console.log libs/sort-package-json/sort-package-json.spec.js:9
{
  unsortedKeys: [
    'name',
    'version',
    'description',
    'author',
    'license',
    'scripts',
    'dependencies',
    'devDependencies',
    'depFn',
    'jest',
    'commitMessage'
  ],
  sortedKeys: [
    'name',
    'scripts',
    'author',
    'version',
    'description',
    'license',
    'dependencies',
    'devDependencies',
    'depFn',
    'jest',
    'commitMessage'
  ]
}

----------------------|----------|----------|----------|----------|-------------------|
File                  |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
----------------------|----------|----------|----------|----------|-------------------|
All files             |    92.31 |       60 |      100 |    91.67 |                   |
sort-package-json.js |    92.31 |       60 |      100 |    91.67 |                29 |
----------------------|----------|----------|----------|----------|-------------------|
`