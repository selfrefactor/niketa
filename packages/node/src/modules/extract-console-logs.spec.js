import {
  testInput,
  testInputSecond,
  testInputThird,
} from '../../test-data/extract-logs/data'
import { newJest } from '../../test-data/extract-logs/new-jest'
import { oldJest } from '../../test-data/extract-logs/old-jest'
import { extractConsoleLogs, withNewJest } from './extract-console-logs'

test.only('new jest', () => {
  const result = withNewJest(newJest)
  expect(result).toMatchSnapshot()
})

test('happy', () => {
  const result = extractConsoleLogs(testInput)
  expect(result).toMatchSnapshot()
})

test('multiple logs on same line', () => {
  const result = extractConsoleLogs(testInputSecond)
  expect(result).toMatchSnapshot()
})

test('bug', () => {
  const result = extractConsoleLogs(testInputThird)
  expect(result).toMatchSnapshot()
})
