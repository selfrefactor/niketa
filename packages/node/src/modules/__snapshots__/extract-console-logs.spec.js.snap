// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new jest 1`] = `
Object {
  "libs/sort-package-json/sort-package-json.spec.js": Array [
    Object {
      "decoration": "{ type: 'object' } { type: 'boolean' } ",
      "lineNumber": "6",
    },
    Object {
      "decoration": "{ array: 'not array' }",
      "lineNumber": "7",
    },
    Object {
      "decoration": "{ keys: [ 'unsortedKeys', 'sortedKeys' ] }",
      "lineNumber": "8",
    },
    Object {
      "decoration": "{ unsortedKeys: [ 'name', 'version', 'description', 'author', 'license', 'scripts', 'dependencies', 'devDependencie",
      "lineNumber": "9",
    },
  ],
}
`;
